neurom/core/types.py

```diff
@@ -47,48 +48,107 @@ class NeuriteIter(OrderedEnum):
     NRN = 2
 
 
-class Subtypes:
+class NeuriteSubtype(int):
+    _ALL = 32
+    _BASE = 100
+    _COMPOSITE = 99
 
-    def __init__(self, *subtypes):
-        self.subtypes = subtypes
+    def __new__(cls, *value):
+        if len(value) == 1:
+            # Avoid recursion error
+            value = value[0]
+
+        if isinstance(value, (list, tuple)):
+            value = cls.from_list(value)
+        elif isinstance(value, Enum):
+            value = value.value
+        obj = super().__new__(cls, value)
+        obj._subtypes = NeuriteSubtype.to_list(obj)
+        obj._value_ = int(obj)
+        return obj
+
+    def __hash__(self):
+        return hash(self._value_)
 
     def is_composite(self):
-        return len(self.subtypes) > 1
+        return self._value_ >= self._BASE
 
     def __eq__(self, other):
-
-        if isinstance(other, Subtypes):
+        if not isinstance(other, NeuriteSubtype):
+            other = NeuriteSubtype(other)
+        if self._ALL in self._subtypes or self._ALL in other._subtypes:
```

Comment:
_ALL should not be a subtype.

Reply:
I was not sure about this but I think it can be interesting because then we don't have to write special cases everywhere in the code to handle the `all` case. With this comparison it will just always be considered as equal to all other subtypes, so I think it can simplify the code of `NeuroM`. But this was just an idea, we can remove it.

Reply:
I would expect to pass an ALL as a type but not as a subtype (e.g. [Axon, All]).
So instead sth like:
```python
if self._ALL in {self._value_, other._value_}:
```

Reply:
So you would also forbid to create subtypes containing the `_ALL` value (e.g. `[2, 32] == [axon, all]`), right?

Reply:
Yes, they don't make much sense tbh because they define a group, not a single mixed type. wdyt?

Reply:
Btw, the current implementation is very flexible:
```python
SubtypeCollection(3, 2, 1) == _ALL_SUBTYPE
SubtypeCollection(3, 2, 1) == SubtypeCollection(_ALL_SUBTYPE)
SubtypeCollection(3, 2, 1) == SubtypeCollection([axon, _ALL_SUBTYPE])
```

Reply:
Well, I agree that defining a subtype containing the `_ALL_SUBTYPE` subtype does not make sense. But I think that forbidding such subtypes is enough, what happens in the `__eq__` operator is just an implementation detail (but the values in the `_subtypes` attributes are always integers, so it's easy to compare, while the values in the `_value_` attribute can be a `SubtypeCollection`, which makes the comparison a bit more cumbersome, though I can change this).